internal/dag/gatewayapi_processor.go

```diff
@@ -230,24 +230,98 @@ func isSecretRef(certificateRef *gatewayapi_v1alpha1.LocalObjectReference) bool
 	return strings.ToLower(certificateRef.Kind) == "secret" && strings.ToLower(certificateRef.Group) == "core"
 }
 
-func (p *GatewayAPIProcessor) computeHosts(hostnames []gatewayapi_v1alpha1.Hostname) ([]string, []error) {
-	// Determine the hosts on the route, if no hosts
-	// are defined, then set to "*".
-	var hosts []string
+// computeHosts validates the hostnames for a HTTPRoute as well as validating
+// that the hostname on the HTTPRoute matches what is optionally defined on the
+// listener.hostname.
+func (p *GatewayAPIProcessor) computeHosts(hostnames []gatewayapi_v1alpha1.Hostname, listenerHostname *gatewayapi_v1alpha1.Hostname) (map[string]struct{}, []error) {
+
+	hosts := make(map[string]struct{})
 	var errors []error
-	if len(hostnames) == 0 {
-		hosts = append(hosts, "*")
+
+	// Determine the hosts on the hostnames, if no hosts
+	// are defined, then set to "*". If the listenerHostname is defined,
+	// then the route must match the Gateway hostname.
+	if len(hostnames) == 0 && listenerHostname == nil {
+		hosts["*"] = struct{}{}
 		return hosts, nil
 	}
 
+	if listenerHostname != nil {
+		if string(*listenerHostname) != "*" {
+
+			// Validate listener hostname.
+			if err := validHostName(string(*listenerHostname)); err != nil {
+				return hosts, []error{err}
+			}
+
+			if len(hostnames) == 0 {
+				hosts[string(*listenerHostname)] = struct{}{}
+				return hosts, nil
+			}
+		}
+	}
+
 	for _, host := range hostnames {
 
 		hostname := string(host)
+
+		// A "*" hostname matches anything.
+		if hostname == "*" {
+			hosts[hostname] = struct{}{}
+			continue
+		}
+
+		// Validate the hostname.
 		if err := validHostName(hostname); err != nil {
 			errors = append(errors, err)
 			continue
 		}
-		hosts = append(hosts, string(host))
+
+		if listenerHostname != nil {
+			const wildcardString string = "[a-z0-9]([-a-z0-9]*[a-z0-9])*."
```

Comment:
just noticed, do we have to escape the `.` at the end b/c we want a literal `.`?

Reply:
took out all the regex stuff with your suggestion.